server/schedule/filter/filters.go

```diff
@@ -492,3 +494,78 @@ func (f *BlacklistStoreFilter) filter(store *core.StoreInfo) bool {
 	_, ok := f.blacklist[store.GetID()]
 	return ok
 }
+
+// labelConstraintFilter is a filter that selects stores satisfy the constraints.
+type labelConstraintFilter struct {
+	scope       string
+	constraints []placement.LabelConstraint
+}
+
+// NewLabelConstaintFilter creates a filter that selects stores satisfy the constraints.
+func NewLabelConstaintFilter(scope string, constraints []placement.LabelConstraint) Filter {
+	return labelConstraintFilter{scope: scope, constraints: constraints}
+}
+
+// Scope returns the scheduler or the checker which the filter acts on.
+func (f labelConstraintFilter) Scope() string {
+	return f.scope
+}
+
+// Type returns the name of the filter.
+func (f labelConstraintFilter) Type() string {
+	return "label-constraint-filter"
+}
+
+// Source filters stores when select them as schedule source.
+func (f labelConstraintFilter) Source(opt opt.Options, store *core.StoreInfo) bool {
+	return !placement.MatchLabelConstraints(store, f.constraints)
+}
+
+// Target filters stores when select them as schedule target.
+func (f labelConstraintFilter) Target(opt opt.Options, store *core.StoreInfo) bool {
+	return !placement.MatchLabelConstraints(store, f.constraints)
+}
+
+// RegionFitter is the interface that can fit a region against placement rules.
+type RegionFitter interface {
+	FitRegion(*core.RegionInfo) *placement.RegionFit
+}
+
+type ruleFitFilter struct {
+	scope     string
+	fitter    RegionFitter
+	region    *core.RegionInfo
+	oldFit    *placement.RegionFit
+	IsLearner bool
+}
+
+// NewRuleFitFilter creates a filter that ensures that the region fitness won't decrease.
+func NewRuleFitFilter(scope string, fitter RegionFitter, region *core.RegionInfo, removePeer *metapb.Peer) *ruleFitFilter {
+	oldFit := fitter.FitRegion(region)
+	region = region.Clone(core.WithRemoveStorePeer(removePeer.GetStoreId()))
```

Comment:
How about record `origin region` and `replace store` to reduce once clone?

Reply:
good idea.